File: services/ai/src/components/replay-memory/replay_memory.ts
import { ReplayMemoryItem } from '@core/types/t_game';
import * as Tf from '@tensorflow/tfjs';
 
export class ReplayMemory {
	
	private	_max_len: number;
	private	buffer: (ReplayMemoryItem | null)[] = new Array();
	private	buffer_index: number[] = new Array();
	private	length: number = 0;
	private	index: number = 0;

	public	get	max_len() {return (this._max_len)}
	public	get	appended() {return (this.length)}

	constructor (max_len: number) {
		this._max_len = max_len;
		for (let i = 0; i < this._max_len; ++i) {
			this.buffer.push(null);
			this.buffer_index.push(i);
		}
	}

	ft_append(item: ReplayMemoryItem) {
		this.buffer[this.index] = item;
		this.length = Math.min(this.length + 1, this._max_len);
		this.index = (this.index + 1) % this._max_len;
	}

	ft_sample(batch_size: number): ReplayMemoryItem[] {
		if (batch_size > this._max_len) {
		  throw new Error(
			  `batchSize (${batch_size}) exceeds buffer length (${this._max_len})`);
		}
		Tf.util.shuffle(this.buffer_index);

		const out: ReplayMemoryItem[] = [];
		for (let i = 0; i < batch_size; ++i) {
			const	rp_item = this.buffer[this.buffer_index[i]!];
			if (rp_item != undefined && rp_item != null)
				out.push(rp_item);
		}
		return out;
	}
}

File: services/ai_python/src_py/components/replay_memory/replay_memory.py
import random
from typing import List, Optional
from src_py.types.t_game import ReplayMemoryItem

class ReplayMemory:
    def __init__(self, max_len: int):
        self._max_len = max_len
        self.buffer: List[Optional[ReplayMemoryItem]] = [None for _ in range(max_len)]
        self.buffer_index: List[int] = list(range(max_len))
        self.length = 0
        self.index = 0

    @property
    def max_len(self) -> int:
        return self._max_len

    @property
    def appended(self) -> int:
        return self.length

    def ft_append(self, item: ReplayMemoryItem):
        self.buffer[self.index] = item
        self.length = min(self.length + 1, self._max_len)
        self.index = (self.index + 1) % self._max_len

    def ft_sample(self, batch_size: int) -> List[ReplayMemoryItem]:
        if batch_size > self._max_len:
            raise ValueError(f"batchSize ({batch_size}) exceeds buffer length ({self._max_len})")
        random.shuffle(self.buffer_index)
        out: List[ReplayMemoryItem] = []
        i = 0
        while len(out) < batch_size and i < len(self.buffer_index):
            item = self.buffer[self.buffer_index[i]]
            if item is not None:
                out.append(item)
            i += 1
        return out
